### ISSUE TITLE
func-names does not report default exported function

### ISSUE THREAD

**Issue by golopot** [eslint/eslint#12194]
**Tell us about your environment**

* **ESLint Version:** 6.3.0
* **Node Version:** 12

**What parser (default, Babel-ESLint, etc.) are you using?**
default

**Please show your full configuration:**
`func-names: [2, "always"]` or
`func-names: [2, "as-needed"]`

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
/* eslint func-names: [2, 'as-needed'] */
// or  'always'

export default function () {}
```
https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IGZ1bmMtbmFtZXM6IFsyLCAnYXMtbmVlZGVkJ10gKi9cbi8vIG9yICAnYWx3YXlzJ1xuXG5leHBvcnQgZGVmYXVsdCBmdW5jdGlvbiAoKSB7fVxuIiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjo5LCJzb3VyY2VUeXBlIjoibW9kdWxlIiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7fSwiZW52Ijp7fX19

**What did you expect to happen?**
Unnamed function to be reported.

**What actually happened? Please include the actual, raw output from ESLint.**
No error is reported.

**Are you willing to submit a pull request to fix this bug?**
Yes please.

**Notes**
The main purpose of func-names is to avoid unhelpful names like "anonymous" popping up in stack trace or profiling chart. In the case of unnamed default export, the function names will be `_default` for babel transformed code, or `default` for node 12 experimental modules. These names are uninformative for debugging. So func-names should report this case.

**Comment by mysticatea** [comment]
Thank you for your report.

I agreed.

In technical, `func-names` rule has checked only function expressions, but `export default function () {}` is a function declaration. I guess that this is an overlook in ES6 updates.

**Comment by mdjermanovic** [comment]
> In technical, `func-names` rule has checked only function expressions, but `export default function () {}` is a function declaration. I guess that this is an overlook in ES6 updates.

Would this be a bug fix or an enhancement, because by the documentation the rule works only with function expressions.

**Comment by mysticatea** [comment]
Indeed, it's a concern that the fix has big impact.

Personally, this is a bug -- I think that the rule intends the naming convention of functions (regardless of expressions/declarations) as we consider the fact that we could not omit function names on function declarations at the time the rule created. I can remember that the AST of `export default function () {}` was `FunctionExpression` at first, but it has been fixed in order to distinguish `export default function f() {}` and `export default (function f() {});`.

If now is 2015, it's a bug fix definitely. but now...?

I'd like to hear the opinion of the rest members of the team.

**Comment by mdjermanovic** [comment]
I think that the following should be also discussed, because the proposed change doesn't treat function declarations in the same way as the rule at the moment treats function expressions when the option is `as-needed`:

> "as-needed" requires function expressions to have a name, if the name cannot be assigned automatically in an ES6 environment

The actual code is explicitly treating `export default (function () {})` as an automatically assigned name [in this line](https://github.com/eslint/eslint/blob/329e295ff95cb49c1ab5ea4f3efbbc220307cfbd/lib/rules/func-names.js#L122), and for example this is not an error:

```js
/*eslint func-names: ["error", "as-needed"]*/

export default (function () {}) // no error
```

but this is an error:

```js
/*eslint func-names: ["error", "always"]*/

export default (function () {}) // error
```

This is probably technically correct by the documentation, though if the rule's purpose is to have meaningful names in the stack trace (is this assumption correct? the rule is in the `Stylistic Issues` category) it makes sense to change this as well.

**Comment by mysticatea** [comment]
Good point. I agree that to report anonymous default exported functions makes sense.

**Comment by mdjermanovic** [comment]
A good thing for this is that `no-unused-vars` doesn't report `export default function foo () {}` although it is an unused variable in the module scope.

**Comment by platinumazure** [comment]
I think it would be best to treat this as a breaking change or create an option, personally- just in case. I can see the case for it being a bug, but I'm concerned about potential large impact.

**Comment by golopot** [comment]
Creating an option - avoids breaking users' CI for now, but adds a vestige that burdens all future users.

**Comment by mdjermanovic** [comment]
This looks more like 2 options?

One to enforce the rule on function declarations, the other to change `as-needed` logic for `export default`.

**Comment by mysticatea** [comment]
I prefer changing the default behavior on 7.0.0. I'm not sure if there are the cases we use the options.

**Comment by kaicataldo** [comment]
Adding the `breaking` label, as it sounds like consensus is to hold off on this until v7.0.0. Feel free to change that if I misunderstood anything!
